pkg/inventory/inventory-client.go

```diff
@@ -24,455 +24,420 @@ import (
-		return fmt.Errorf("attempting delete a nil inventory object")
-	}
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run delete inventory object: not deleted")
+// CreateOrUpdate the in-cluster inventory
+// Updates the unstructured object, or creates it if it doesn't exist
+func (cic *UnstructuredClient) CreateOrUpdate(ctx context.Context, inv Inventory, opts UpdateOptions) error {
+	ui, ok := inv.(*UnstructuredInventory)
+	if !ok {
+		return fmt.Errorf("expected UnstructuredInventory")
+	}
+	if ui == nil {
+		return fmt.Errorf("inventory is nil")
+	}
+	// Attempt to retry on a resource conflict error to avoid needing to retry the
+	// entire Apply/Destroy when there's a transient conflict.
+	attempt := 0
+	var obj *unstructured.Unstructured
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		defer func() {
+			attempt++
+		}()
+		create := false
+		var err error
+		obj, err = cic.client.Namespace(ui.GetNamespace()).Get(ctx, ui.GetName(), metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			create = true
+			obj = cic.newUnstructuredObject(&ui.SingleObjectInfo)
+		} else if err != nil {
+			return err
+		}
+		uObj, err := cic.toUnstructured(obj.DeepCopy(), ui)
+		if err != nil {
+			return err
+		}
+		if create {
+			klog.V(4).Infof("[attempt %d] creating inventory object %s/%s/%s",
+				attempt, cic.gvk, uObj.GetNamespace(), uObj.GetName())
+			obj, err = cic.client.Namespace(uObj.GetNamespace()).Create(ctx, uObj, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+		} else {
+			klog.V(4).Infof("[attempt %d] updating inventory object %s/%s/%s",
+				attempt, cic.gvk, uObj.GetNamespace(), uObj.GetName())
+			obj, err = cic.client.Namespace(uObj.GetNamespace()).Update(ctx, uObj, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
 		return nil
-	}
-
-	mapping, err := cic.getMapping(obj)
+	})
 	if err != nil {
 		return err
 	}
-
-	klog.V(4).Infof("deleting inventory object: %s/%s", obj.GetNamespace(), obj.GetName())
-	return cic.dc.Resource(mapping.Resource).Namespace(obj.GetNamespace()).
-		Delete(ctx, obj.GetName(), metav1.DeleteOptions{})
-}
-
-// getMapping returns the RESTMapping for the provided resource.
-func (cic *ClusterClient) getMapping(obj *unstructured.Unstructured) (*meta.RESTMapping, error) {
-	return cic.mapper.RESTMapping(obj.GroupVersionKind().GroupKind(), obj.GroupVersionKind().Version)
+	if cic.statusPolicy != StatusPolicyAll || opts.SkipStatusUpdate {
+		return nil
+	}
+	attempt = 0
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		defer func() {
+			attempt++
+		}()
+		if attempt > 0 { // reuse object from spec update/create on first attempt
+			obj, err = cic.client.Namespace(ui.GetNamespace()).Get(ctx, ui.GetName(), metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		uObj, err := cic.toUnstructured(obj.DeepCopy(), ui)
+		if err != nil {
+			return err
+		}
+		// Update observedGeneration, if it exists
```

Comment:
Add a comment that says something like "If using a custom inventory resource, make sure the observedGeneration defaults to 0, so that it will be updated here, and so kstatus computes the status as InProgress after the object is created, until the status is updated here."

Reply:
Done